### PR TITLE
Feature exec

### PR DESCRIPTION
**Pull request by gaimanjing**
Implemented: vgg exec.

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -14,14 +14,41 @@ add_executable(runtime
 )
 
 add_library(vgg_internal OBJECT
+  Exec/vggExec.cpp
```
Comment (harry75369) [inline src/CMakeLists.txt:17]: Failed to run cmake under linux. Should be `VggExec.cpp`

File: src/PlatformAdapter/Native/Exec/NativeExecImpl.cpp [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+ */
+bool NativeExecImpl::schedule_eval(const std::string& code)
+{
+  if (m_state == DEAD)
+  {
+    FAILED("#NativeExecImpl::schedule_eval, error, dead state");
+    return false;
+  }
+
+  while (m_state != RUNNING)
```
Comment (harry75369) [inline src/PlatformAdapter/Native/Exec/NativeExecImpl.cpp:22]: should add doc to explain m_state will be modified somewhere, otherwise it'll be spinning forever.

File: src/PlatformAdapter/Native/Exec/NativeExecImpl.cpp [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+  {
+  }
+
+  NativeEvalTask* task = new NativeEvalTask();
+  task->m_code = code;
+  task->m_exec_impl_ptr = this;
+
+  const std::lock_guard<std::mutex> lock(m_tasks_mutex);
+  m_tasks.insert(task);
```
Comment (harry75369) [inline src/PlatformAdapter/Native/Exec/NativeExecImpl.cpp:31]: could this mutex lock be released immediately after insert? Releasing it after uv invokes may be not good.

File: src/PlatformAdapter/Native/Exec/NativeExecImpl.cpp [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+
+  uv_async_init(m_loop,
+                &task->m_async_task,
+                [](uv_async_t* async)
+                {
+                  uv_close(reinterpret_cast<uv_handle_t*>(async), nullptr);
+                  NativeEvalTask* task = node::ContainerOf(&NativeEvalTask::m_async_task, async);
+                  task->run();
+                });
+  uv_async_send(&task->m_async_task);
```
Comment (harry75369) [inline src/PlatformAdapter/Native/Exec/NativeExecImpl.cpp:41]: add doc to explain these two invokes, otherwise those who don't know libuv will get confused.

File: src/PlatformAdapter/Native/Exec/NativeExecImpl.cpp [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+
+  const std::lock_guard<std::mutex> lock(m_tasks_mutex);
+  m_tasks.insert(task);
+
+  uv_async_init(m_loop,
+                &task->m_async_task,
+                [](uv_async_t* async)
+                {
+                  uv_close(reinterpret_cast<uv_handle_t*>(async), nullptr);
+                  NativeEvalTask* task = node::ContainerOf(&NativeEvalTask::m_async_task, async);
```
Comment (harry75369) [inline src/PlatformAdapter/Native/Exec/NativeExecImpl.cpp:38]: why is this line necessary?

File: src/PlatformAdapter/Native/Exec/NativeExecImpl.cpp [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+      node::ProcessInitializationFlags::kNoInitializeNodeV8Platform });
+
+  for (const std::string& error : result->errors())
+    fprintf(stderr, "%s: %s\n", args[0].c_str(), error.c_str());
+  if (result->early_return() != 0)
+  {
+    return result->exit_code();
+  }
+
+  std::unique_ptr<MultiIsolatePlatform> platform = MultiIsolatePlatform::Create(4);
```
Comment (harry75369) [inline src/PlatformAdapter/Native/Exec/NativeExecImpl.cpp:115]: should add doc to explain `Create(4)`

File: src/PlatformAdapter/Native/Exec/NativeExecImpl.cpp [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+int NativeExecImpl::run_node(const int argc,
+                             const char** argv,
+                             std::shared_ptr<std::thread>& nodeThread,
+                             const std::function<void(const node::Environment*)>& envHook,
+                             const std::function<const char*()>& getScriptHook)
+{
+  uv_setup_args(argc, const_cast<char**>(argv));
+  std::vector<std::string> args(argv, argv + argc);
+
+  nodeThread.reset(new std::thread([&, args]() { node_main(args, envHook, getScriptHook); }));
```
Comment (harry75369) [inline src/PlatformAdapter/Native/Exec/NativeExecImpl.cpp:94]: should we care about the return code of `node_main`?

File: src/PlatformAdapter/Native/Exec/NativeExecImpl.cpp [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+                  uv_close(reinterpret_cast<uv_handle_t*>(&self_ptr->m_stop_timer_async), nullptr);
+                });
+  m_stop_timer_async.data = static_cast<void*>(this);
+  uv_async_send(&m_stop_timer_async);
+}
+
+void NativeExecImpl::run_task(NativeEvalTask* task)
+{
+  DEBUG("#evalScript, before eval");
+  eval(task->m_code);
```
Comment (harry75369) [inline src/PlatformAdapter/Native/Exec/NativeExecImpl.cpp:241]: should we care about the return code of `eval`? `schedual_eval` always returns true, so we cannot get the actual eval result.

File: src/PlatformAdapter/Native/Exec/NativeExecImpl.cpp (hunk 3 above)
Comment (gaimanjing) [reply src/PlatformAdapter/Native/Exec/NativeExecImpl.cpp:31]: Yes, updated.

File: src/PlatformAdapter/Native/Exec/NativeExecImpl.cpp (hunk 4 above)
Comment (gaimanjing) [reply src/PlatformAdapter/Native/Exec/NativeExecImpl.cpp:41]: added

File: src/PlatformAdapter/Native/Exec/NativeExecImpl.cpp (hunk 5 above)
Comment (gaimanjing) [reply src/PlatformAdapter/Native/Exec/NativeExecImpl.cpp:38]: compile error when the lambda passed to uv_async_init captures variable.

File: src/PlatformAdapter/Native/Exec/NativeExecImpl.cpp (hunk 6 above)
Comment (gaimanjing) [reply src/PlatformAdapter/Native/Exec/NativeExecImpl.cpp:115]: added

File: src/PlatformAdapter/Native/Exec/NativeExecImpl.cpp (hunk 7 above)
Comment (gaimanjing) [reply src/PlatformAdapter/Native/Exec/NativeExecImpl.cpp:94]: add log for debug.

File: src/PlatformAdapter/Native/Exec/NativeExecImpl.cpp (hunk 8 above)
Comment (gaimanjing) [reply src/PlatformAdapter/Native/Exec/NativeExecImpl.cpp:241]: add log to print the ret code, and there are logs in the `eval` method.
